node/libs/roles/src/validator/messages/leader_proposal.rs

```diff
@@ -0,0 +1,208 @@
+        }
+
+        if implied_payload.is_some() && self.proposal_payload.is_some() {
+            return Err(LeaderProposalVerifyError::NewProposalWhenPreviousNotFinalized);
+        }
+
+        // Check that the payload matches the header, if it exists.
+        if let Some(payload) = &self.proposal_payload {
+            if payload.hash() != self.proposal.payload {
+                return Err(LeaderProposalVerifyError::MismatchedPayload {
+                    header: self.proposal.payload,
+                    payload: payload.hash(),
+                });
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Error returned by `LeaderProposal::verify()`.
+#[derive(thiserror::Error, Debug)]
+pub enum LeaderProposalVerifyError {
+    /// Invalid Justification.
+    #[error("Invalid justification: {0:#}")]
+    Justification(ProposalJustificationVerifyError),
+    /// Bad block number.
+    #[error("Bad block number: got {got:?}, want {want:?}")]
+    BadBlockNumber {
+        /// Received proposal number.
+        got: BlockNumber,
+        /// Correct proposal number.
+        want: BlockNumber,
+    },
+    /// Bad payload hash on reproposal.
+    #[error("Bad payload hash on reproposal: got {got:?}, want {want:?}")]
+    BadPayloadHash {
+        /// Received payload hash.
+        got: PayloadHash,
+        /// Correct payload hash.
+        want: PayloadHash,
+    },
+    /// New block proposal when the previous proposal was not finalized.
+    #[error("New block proposal when the previous proposal was not finalized")]
+    NewProposalWhenPreviousNotFinalized,
+    /// Re-proposal when the previous proposal was finalized.
+    #[error("Block re-proposal when the previous proposal was finalized")]
+    ReproposalWhenPreviousFinalized,
+    /// Mismatched payload.
+    #[error("Block proposal with mismatched payload: header {header:?}, payload {payload:?}")]
+    MismatchedPayload {
+        /// Payload hash on block header.
+        header: PayloadHash,
+        /// Correct payload hash.
+        payload: PayloadHash,
+    },
+}
+
+/// Justification for a proposal. This is either a Commit QC or a Timeout QC.
+/// The first proposal, for view 0, will always be a timeout.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ProposalJustification {
+    /// This proposal is being proposed after a view where we finalized a block.
+    /// A commit QC is just a collection of commit votes (with at least
+    /// QUORUM_WEIGHT) for the previous view. Note that the commit votes MUST
+    /// be identical.
+    Commit(CommitQC),
+    /// This proposal is being proposed after a view where we timed out.
+    /// A timeout QC is just a collection of timeout votes (with at least
+    /// QUORUM_WEIGHT) for the previous view. Unlike with the Commit QC,
+    /// timeout votes don't need to be identical.
+    /// The first proposal, for view 0, will always be a timeout.
```

Comment:
I think it is in fact a good idea to make view 0 always time out, as this will be the initial synchronization point of the consensus. However please make sure that this doesn't create an artificial delay - i.e. validators should immediately cast timeout vote for view 0 on startup, rather than waiting for the actual timeout delay.